### ISSUE TITLE
Missing chat messages notifications

### ISSUE THREAD

**Issue by e-caste** [nextcloud/talk-ios#595]
## Steps to reproduce
1. Send a private message to another user
2. The push notification may or may not arrive. I have yet to identify the cause, this issue aims to solve this problem.

### Expected behaviour
A push notification should show up immediately and 100% of the time if a message is received.

### Actual behaviour
That usually does not happen. The message is received, but until I open the app I can't see if there is a new message or not.  
I have double checked all my settings: the Talk chat notifications are enabled, the iOS notifications for this app are enabled, I have solved all warnings in the Nextcloud Overview settings section.  
As a note: notifications used to work perfectly on v10. With v11, some notifications started to be missing, while sometimes they were received correctly. Now with v12 I almost never receive them. I have tried reinstalling the app from the App Store, and that seemed to work for a couple hours (meaning I was correctly receiving all notifications immediately), but that solution seems to be very temporary.

As a sidenote, my setup has 2 "peculiarities" that might trigger this wrong behavior:
- my Nextcloud data is on an NFS mount passed to Docker
- on my private network I use a local DNS (though I experience the same behavior both when connected to it and when I'm outside)  

As a final note, my Nextcloud instance is behind a reverse proxy, and although I believe that is the setup most people have, I might have a missing or wrong line here and there in the config file (but as I pointed out before, I have no warnings in my Overview settings section).

EDIT: it seems that I receive notifications for replies to my messages, but not for other direct messages (I've reproduced this behavior three times, quite sure this is one of the problems). I have double checked with the web version and the chat notifications are set to "all messages". Could this be an issue with nextcloud/spreed instead of the mobile app?  
I have also tried with another private chat and that seems to work fine. I can't pinpoint the issue, any help is welcome.

EDIT 2: after a video call yesterday I started receiving all notifications correctly. I don't know how long it will last, but this is my current situation. I have not made any configuration changes or updates. Even so, I would still like to know why notifications haven't worked for a few days, and if this is possible to avoid in the future, so that I can recommend this platform to more people. Let me know if any more detail is needed to debug what happened.

EDIT 3: amending EDIT 2, now notifications are actually hit and miss on both ends of the conversation.

## Device information

**Device:** iPhone 7, iPhone SE 1

**iOS version:** both 14.6

**Talk version:** both 12.1.0

## Server information

**Nextcloud version:** (see admin overview page: `/index.php/settings/admin/overview`) 22.0.0

**Talk version:** (see apps admin page: `/index.php/settings/apps`) 12.0.1

**Custom Signaling server configured:** yes/no and version (see talk admin settings: `/index.php/settings/admin/talk#signaling_server`) no

**Custom TURN server configured:** yes/no (see talk admin settings: `/index.php/settings/admin/talk#turn_server`) yes

**Custom STUN server configured:** yes/no (see talk admin settings: `/index.php/settings/admin/talk#stun_server`) no (default STUN server)

### Server log (data/nextcloud.log)
<details>

```
Insert your server log here
```
</details>

**Comment by chosebine501** [comment]
same issues
not icon, no notfication after last update

with version 12.0 no icon but the notification it's ok
now nothing left…

**Comment by chosebine501** [comment]
It's important to know that in my environment we use Talk to chat, only that on IOS (Iphone and Ipad) we never use web browsers.
After updating to version 12 of Talk, no more notifications work. for anyone.

For us, the notifications between IOS started to work again, after I connected with a web browser and sent a message to each of my contacts, and the notifications started to work again for everyone

**Comment by chosebine501** [comment]
> It's important to know that in my environment we use Talk to chat, only that on IOS (Iphone and Ipad) we never use web browsers.
> After updating to version 12 of Talk, no more notifications work. for anyone.
> 
> For us, the notifications between IOS started to work again, after I connected with a web browser and sent a message to each of my contacts, and the notifications started to work again for everyone

After multi test
at least once or twice a day, I have to reconnect with the web page and resend a message, for the notifications to work again between IOS devices.

**Comment by e-caste** [comment]
> After multi test
> at least once or twice a day, I have to reconnect with the web page and resend a message, for the notifications to work again between IOS devices.

Unfortunately this does not seem to work for me (even after sending a message from Nextcloud web). I still only get notifications for replies to my messages, but not for direct messages.

**Comment by chosebine501** [comment]
Hi,
dev talk team, do you think you will have a definitive solution soon.
I'm asking you because for my part, there is nobody who wants to use it anymore, I try to make them wait, but unfortunately, in our world nobody wants to wait anymore...
so I'm the only one who still wants to use it, but alone is not very useful 
regars

**Comment by staticvoidmaine** [comment]
Hi there,

I'm just dropping another data point that as an iOS and Web Talk user, notifications are at best flaky, and at worst never arrive in recent versions.

**Comment by ChildLearningClub** [comment]
just wanted to also link this here from the forum for reference [Nextcloud Talk push notifications not working](https://help.nextcloud.com/t/nextcloud-talk-push-notifications-not-working/120873)

**Comment by AndyXheli** [comment]
Yes I can confirm this I'm not getting notifications

**Comment by andyhubbert941** [comment]
Hi
same issue here where notifications for a private message thread are intermittent on both ends of the conversation. Maybe an issue with background app refresh? I’m not a dev though so no real idea!

Nextcloud configured as per install guide and no other issues with my server. I am not using docker and have it installed bare metal so the install doesn’t appear to matter

**Comment by e-caste** [comment]
@andyhubbert941 Yes, my issue with the latest versions of all, updated with respect to my original submission (now Nextcloud 22.1.1, Talk Server 12.1.1, Talk iOS 12.2.0) is the same as yours: _intermittent_ notifications.  
I think it correlates with Nextcloud Talk video calls in some way, meaning that after a call the party that was receiving notifications stops receiving them and viceversa, but not completely, i.e. both parties also randomly start and stop receiving notifications.  

### To add to my original post:  

My server installation is based on docker-compose, with the following images involved:
- postgres:alpine (latest)
- nextcloud:fpm-alpine (customized with [this Dockerfile](https://github.com/nextcloud/docker/blob/master/.examples/dockerfiles/full/fpm-alpine/Dockerfile))
- nginx:alpine (latest, with nginx.conf taken from the documentation)
- redis (latest)

My storage is configured as following:
- base files in Docker volume on SSD
- data directory in Docker volume on NFS mount (maybe this could be an issue? It seems to work reliably in all the other situations)

**Comment by chosebine501** [comment]
It's been 4 months…

**Comment by andyhubbert941** [comment]
Yes but as we are both still experiencing the issue then it is relevant and any help would be appreciated


On 9 Sep 2021, at 13:51, chosebine501 ***@***.***> wrote:

﻿

It's been 4 months…

**Comment by AndyXheli** [comment]
Having this same issue no Call or message notifications This really needs to be looked at it makes the app useless i call my users and they dont know im calling i have to text them to join the call whats the point of the app

**Comment by e-caste** [comment]
@Ivansss Hi, I know it's somewhat rude to ask you directly, but I think at this point I (and all those who have had this issue as well, who are following it) need to know: what is the priority that the Nextcloud organization has assigned to this issue? It seems to be widespread, and to me and to other people it ruins a user experience that would instead be very good if it was resolved. Will we have to wait for the next minor/major Nextcloud or Spreed release? Is it going to be worked on at some point?  
Thank you in advance, sorry for bothering you!

**Comment by Ivansss** [comment]
Hello @e-caste,
No worries for the ping and thank you for your detailed posts in this issue :)

Unfortunately we can't reproduce this issue.
We fixed a bug that caused delayed push notifications in version 12.1.0.
We still experience from time to time delays in push notifications but we think that they could be produced by iOS itself (due to receiving many push notification in a short period of time)

We take this topic (push notifications) really seriously as we know is the main functionality of a messenger app.
But we need a reliable way to reproduce this issue in order to help the people who is experiencing this issue.

In order to test if notifications correctly work in your setup you can try with the following occ command:
```
occ notification:test-push --talk USERID
```

I saw your post here https://github.com/nextcloud/spreed/issues/6096#issuecomment-916838082

If you receive a push notification in the Talk iOS app, it means that the device is correctly subscribed to the push notification proxy server.

One common reason that causes a chat notification not being sent is when the user has another session opened in that conversation in another device (e.g. in the web)

For those users who still experience missing push notifications, I would recommend to revoke all not-used Nextcloud Talk sessions in Settings -> Security -> Devices & sessions (using the Web interface), then re-install the Talk iOS app in that device and log-in again.

I please ask anyone who is experiencing this issue to provide detailed information about "how to reproduce this issue" or "in which situations it seems to happen" instead of posting that they have the same issue (please react with a thumb-up to the original post of this issue if you are experiencing this issue).

**Comment by Ivansss** [comment]
For the ones that has been experiencing missing call notifications these last days, there was an issue in our push proxy that has been fixed today (ref: https://github.com/nextcloud/talk-ios/issues/627)
This issue only affected call notifications but chat notifications were not affected.

**Comment by SystemKeeper** [comment]
> One common reason that causes a chat notification not being sent is when the user has another session opened in that conversation in another device (e.g. in the web)

Maybe a log output (debug) would be useful and could be added in such case, so it's easier to verify that this is what indeed happens? I guess it should be in https://github.com/nextcloud/spreed/blob/b9be9ae1b1dba86f7c7dbec927438800cd3636f0/lib/Chat/Notifier.php#L371-L392 ? Might be a bit noisy, though.

**Comment by AndyXheli** [comment]
@Ivansss would this fix https://github.com/nextcloud/talk-ios/issues/582 ? and is there anything we need to update on our servers to fix this push proxy issuse

**Comment by andyxheli** [comment]
Some of my users get it and some do not I dont know if IOS 15 as anything to do with it but i know this users that having the issue was not having it before ios 15

**Comment by AndyXheli** [comment]
i dont know if this is related but on IOS 15 they have an option called Focus and if you enable one of those options and talk is not in the allowed app it will stop working.  I has my users that are using IOS 15 add Nextcloud talk under the allowed apps. 

If this is off topic please let me know and i can mark it off topic or delete it. 

![image](https://user-images.githubusercontent.com/59488153/135159845-dd6a4744-d767-4d92-9e63-7873b8b99b96.png)

**Comment by SystemKeeper** [comment]
@andyxheli Focus is a new feature in iOS 15 which needs to be turned on explicitly. Much like "do not disturb" in older iOS versions, but with multiple profiles and much more fine grained control about what is allowed for each profile. If your users enable a "focus profile" then Talk needs to be added, as it is with every other app.

When an app receives a notification which is not in the allowed list, it will be displayed in notificiation center, something like "This happend during 'do not disturb'... Nextcloud Talk". So the notification should still be there, just not as an alert (of course notifications need to be turned on for talk in general):
![image](https://user-images.githubusercontent.com/1580193/135162894-144c3eec-aca3-40b2-8de6-d75b4b9369d2.png)

If you think push notifications in general are not working (iOS 15 or not), try the command which Ivan posted above. This tests if push in general is working for this user.

**Comment by AndyXheli** [comment]
Yes I agree with you personally I only use it at night when I'm sleeping and turn it off but for the people that like it on good idea to add the app if they wanna get notifications. As of right now everything seems to be working for me I get call notifications and messages

**Comment by SystemKeeper** [comment]
Sure, but I guess in this case we can agree that we are definitely not talking about a problem with nextcloud or talk-ios 😉 And as said, this topic is about missing notifications in general, not the ones which are not raised because of the settings of a user. So one way or another, it boils down to a reproducable scenario, so it can be addressed.

**Comment by e-caste** [comment]
@andyxheli I think you're missing the point of this issue: it was opened because, with default settings, all notifications enabled, in normal conditions, and already on iOS 14 (but still on 15), many people and I do not receive push notifications for chat messages.  
It is very normal and to be expected that you will not receive notifications for an app which you have actively muted, with Focus, Do Not Disturb, or any other setting.  
This is outside the scope of the issue, and not pertinent.

**Comment by AndyXheli** [comment]
I'm going to mark my comments as of topic. :) sorry for the confusion.

**Comment by e-caste** [comment]
@Ivansss thank you for taking the time to reply.  
> Unfortunately we can't reproduce this issue.  

I would love to help you find the difference between our environments: how would I do that most efficiently? Should I look for something specific in the server logs, or a possible misconfiguration in the database? I have migrated it from a nextcloud/vm to a nextcloud/docker installation, so maybe that may have produced some inconsistencies.
<hr/>

> We still experience from time to time delays in push notifications

This is weird: when I get push notifications, they are usually instantaneous.  
My issue is different: I actually do not get any notification at all, most of the time.
<hr/>

> One common reason that causes a chat notification not being sent is when the user has another session opened in that conversation in another device (e.g. in the web)

I suspect this may be what is happening: I usually have my Safari desktop open, with a tab on my Nextcloud page. After a while, it should perform an automatic logout and delete the cookie, and it seems to work.  
Instead, I have had the opposite problem: the iOS app does not always correctly terminate a video call, and only when I log in the browser the call is terminated. I believe being in a video call affects the push notifications, though I am no Nextcloud dev. Could you please look into this?
<hr/>

> For those users who still experience missing push notifications, I would recommend to revoke all not-used Nextcloud Talk sessions in Settings -> Security -> Devices & sessions (using the Web interface), then re-install the Talk iOS app in that device and log-in again.

I have done that a couple of times now, but it doesn't seem to affect the push notifications. I have experimented with the `occ` command and found that after logging out old sessions from Nextcloud and Nextcloud Talk the server tries to send notifications to fewer devices, but this has not fixed my problem.
<hr/>

And finally, I forgot to add: I **always** receive push notifications for replies to my messages, even in 1-on-1 chats. The issue is with the messages that are not direct replies. (I have not tested with mentions)

**Comment by SystemKeeper** [comment]
> And finally, I forgot to add: I **always** receive push notifications for replies to my messages, even in 1-on-1 chats. The issue is with the messages that are not direct replies. (I have not tested with mentions)

If I'm not mistaken this does sound like an active session. When a user is mentioned, there's no check if the user still has an active session or not (see https://github.com/nextcloud/spreed/blob/b9be9ae1b1dba86f7c7dbec927438800cd3636f0/lib/Chat/Notifier.php#L313-L356). In contrast to a 'normal' notification: https://github.com/nextcloud/spreed/blob/b9be9ae1b1dba86f7c7dbec927438800cd3636f0/lib/Chat/Notifier.php#L371-L392

Just to make sure about

> the iOS app does not always correctly terminate a video call,

There's no way you can terminate the video call with talk-ios? Like leaving the room and joining again or the like?

**Comment by Ivansss** [comment]
> If I'm not mistaken this does sound like an active session. When a user is mentioned, there's no check if the user still has an active session or not 

You are totally right here @SystemKeeper :)
If you (@e-caste) always receive push notifications in mentions or replies, then the problem is related with an active session in the conversation where you expect to receive a notification from.

> the iOS app does not always correctly terminate a video call, and only when I log in the browser the call is terminated. I believe being in a video call affects the push notifications

That will definitely affect push notifications, because you will have an active session in that room until you leave the call and the room. So you won't receive chat notifications from that room until then.

Do you know in which scenarios the iOS app can't terminate the call correctly? Can you reproduce that issue?

**Comment by SystemKeeper** [comment]
@Ivansss I assume it's this issue: https://github.com/nextcloud/talk-ios/issues/602

**Comment by AndyXheli** [comment]
@Ivansss is there an issue with push proxy again ? The call notifications stopped working. I'm no logger getting the call notifications running. Is anyone else having this issue ?

**Comment by chosebine501** [comment]
we are many people who have had this problem for 4 months and nothing is done to solve the problem.

**Comment by BJKle** [comment]
I‘ve convinced all my friends to use NC instead of another messenger and created accounts for them. Only Android users now respond, because they get a notification. All IOS users are left out. Regardless of 1to1 or group messages. We spend hours. Tried many different settings. Nothing. 
**This issue really needs to be fixed soon.**

**Comment by e-caste** [comment]
@BJKle as an iOS user I can confirm that I don't receive notifications yet. But that's just because I can't update the Talk app on the server: this issue should be fixed with https://github.com/nextcloud/spreed/pull/6300, which has already been merged and released in the [Talk v12.2.1 pre-release](https://github.com/nextcloud/spreed/releases/tag/v12.2.1) 24 days ago (you can see it backported there as https://github.com/nextcloud/spreed/pull/6329).

@Ivansss do you know why is v12.2.1 in pre-release state and approximately when we will see it in the Nextcloud App Store? I agree with @BJKle that this issue should be fixed ASAP, I want to invite other people to join my Nextcloud Talk chats, but if I can't see the notifications it's pointless.

**Comment by jakobroehrl** [comment]
I could fix the notification problems v12.2.1 on Android and iOS

**Comment by BJKle** [comment]
@Ivansss Will there be an updated version of "Spreed" available for NC22. Version 13 is available for NC23 only.
NC22 ist still on 12.1.2, without the fix.
Thanks

**Comment by e-caste** [comment]
Closing this issue since it is (finally! 🚀) fixed for me with:
- Nextcloud Talk Server v13.0.0
- Nextcloud Talk iOS v13.0.0

**Comment by Ivansss** [comment]
> Will there be an updated version of "Spreed" available for NC22. Version 13 is available for NC23 only.
> NC22 ist still on 12.1.2, without the fix.

There will be a release for NC22 which will include the bugfix as well

**Comment by BJKle** [comment]
> > Will there be an updated version of "Spreed" available for NC22. Version 13 is available for NC23 only.
> > NC22 ist still on 12.1.2, without the fix.
> 
> There will be a release for NC22 which will include the bugfix as well

@Ivansss Is there a rough estimate when it will be available?

**Comment by SystemKeeper** [comment]
> @Ivansss Is there a rough estimate when it will be available?

I guess there is: https://github.com/nextcloud/spreed/releases/tag/v12.2.2 😄

**Comment by Ivansss** [comment]
@BJKle as @SystemKeeper mention v12.2.2 has just been released :)
We tried to release it yesterday but failed uploading it https://github.com/nextcloud/spreed/issues/6666#issuecomment-988084828

**Comment by BJKle** [comment]
It's working. Thank you!!!

**Comment by Entmatix** [comment]
Same issue here with talk 14.0.3 on both IOS (16.0.2), IpadOS (latest) and android (12) across ~10 devices. No notifications, no push. Logout/login in browser/phone app doesn't help. To find there are new messages I have to open application and update chats manually.

**Comment by SystemKeeper** [comment]
Hey @Entmatix this doesn’t sound like something talk-ios specific then. Please check the document at https://github.com/nextcloud/talk-ios/blob/master/docs/notifications.md (there’s one in the android repo as well) and try to get to the bottom of this. If you think it’s a bug in talk-ios, please open a new issue.

**Comment by Entmatix** [comment]
Hello @SystemKeeper !
Thanks for the link.

* Check iOS settings: no "do not disturb", notifications on everywhere, tried to enable/disable, way more than 24 h passed.
* App settings: notifications turned on globally for every message, not only when you are mentioned, in chats notifications are set for all messages, no sessions from browser.
* **App sessions:** tons of duplicates items. I have no idea what it means, but I see over a hundred of duplicates in my user account for devices that are mine and not mine. How is it possible? Dates of last activity started from minutes ago to month ago. Also I'm unable to remove them all (no selection tool or something). I don't think it's a password leak or something of that kind (I use reverse proxy with https, recommended mozilla apache server settings, strong password policy for everything, nobody has access for my home server). Anyway I changed password and going to remove all sessions. For my specific user I have nextcloud talk installed and opened only on my iphone with ios 16.0.2. However I see sessions from ios 15.7, 12.6, android devices (I don't have any). This user is NOT administrator of NC instance. It's unprivileged user for talk and cloud.
* Server settings: I got test notification.
* Push-proxy: it can be reached.

So it seems that my problem is connected to somehow appeared insane amount of duplications of devices. How could I avoid duplications, how could I end all sessions at once, how to explain sessions from devices I've never had? Thanks.
